### ISSUE TITLE
Tracking Issue for rust-lang/rfcs#3681

### ISSUE THREAD

**Issue by estebank** [rust-lang/rust#132162]
This is a tracking issue for the RFC "3681" (rust-lang/rfcs#3681).
The feature gate for the issue is `#![feature(default_field_values)]`.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.
Discussion comments will get marked as off-topic or deleted.
Repeated discussions on the tracking issue may lead to the tracking issue getting locked.

### Steps

- [ ] Implement the RFC (cc @rust-lang/compiler)
    - [ ] Implement the feature
    - [ ] Implement lints described in RFC
- [ ] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- [ ] Formatting for new syntax has been added to the [Style Guide] ([nightly-style-procedure])
- [ ] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs
[nightly-style-procedure]: https://github.com/rust-lang/style-team/blob/master/nightly-style-procedure.md 
[Style Guide]: https://github.com/rust-lang/rust/tree/master/src/doc/style-guide

### Unresolved Questions

XXX --- list all the "unresolved questions" found in the RFC to ensure they are
not forgotten

### Implementation history

- Initial implementation: https://github.com/rust-lang/rust/pull/129514

**Comment by Phosphorus-M** [comment]
Hi @estebank , I have a question 

Are we considering a syntax like the following?

```rust
#[derive(Default)]
struct Pet {
    name: Option<String>, // impl Default for Pet will use Default::default() for name
    age: i128 = 42, // impl Default for Pet will use the literal 42 for age
    foo: Foo = Foo { something: 10 },
}

#[derive(Default)]
struct Foo {
    something: i32 = 2,
}
```

Something like struct inception as default or overwriting the default implementation. 

It's just a question, I don't have the intention to annoying, the RFC is really cool!

**Comment by estebank** [comment]
@Phosphorus-M that would indeed work, `<Pet as Default>::default().name` would be `None` and `<Pet as Default>::default().foo.something` would be `10`. You just wouldn't be able to write `Pet { .. }` without specifying `Pet { name: None, .. }`. Also, the restriction that the default is a `const` might restrict what you can express (like initializing an `Arc`?), but your example would work as expected.

**Comment by SUPERCILEX** [comment]
I couldn't find anything in the RFC about partial defaults... would this real-world example be supported:
```rust
#[derive(Debug)]
struct App {
    epoll: OwnedFd, // ERROR: this is the only field without a default implementation

    manager: Option<AutoDestroy<ZwlrDataControlManagerV1>>,
    seats: Seats,
    pending_offers: PendingOffers,

    error: Option<CliError>,
    pending_seats: SmallVec<(u32, AutoDestroy<WlSeat>), 1>,
}

// It would be really cool if this worked by making the .. use default impls for all the other fields
let app = App { epoll, .. };
```

**Comment by jhpratt** [comment]
@SUPERCILEX See [this section](https://github.com/rust-lang/rfcs/blob/master/text/3681-default-field-values.md#more-fields). It will be supported once the implementation is merged.

**Comment by SUPERCILEX** [comment]
Amazing, thank you and sorry for missing that!

**Comment by SpiderUnderUrBed** [comment]
This is resolved thanks to:
https://github.com/rust-lang/rust/pull/129514

**Comment by estebank** [comment]
@SpiderUnderUrBed It can't be resolved until the feature gets stabilized. This ticket is to track *all* the work, from initial nightly-only implementation up until it being usable in stable, with anything in between (like checkmarks that are still unmarked in the description).

**Comment by estebank** [comment]
@Phosphorus-M I created https://github.com/rust-lang/rust/pull/135859 to lint against [the case you presented earlier](https://github.com/rust-lang/rust/issues/132162#issuecomment-2488775495). Doing that by accident can lead to confusion, but making it a lint allows for APIs that need it to still be able to enable it and do it.

**Comment by vE5li** [comment]
I have a question regarding generics and I hope it has not been asked before (I searched but couldn't find anything).

Would something like this
```rust
struct Foo<T> {
    bar: bool,
    baz: T = false,
}
```
be allowed?

I don't know if this is even possible but it would be very useful to work with dynamic dispatch and generic types.

E.g.:
```rust
let mut foos: Vec<Box<dyn Any>> = Vec::new();

foos.push(Box::new(Foo { bar: true, .. }));
foos.push(Box::new(Foo { bar: false, baz: 10u64 }));
```

Keep up the great work! :)

**Comment by jhpratt** [comment]
@vE5li Definitely not. The initializer should be thought of as `const baz: T = false;` for the purposes of type checking.

**Comment by Phosphorus-M** [comment]
Hi, I'm here again with a new question! 
The default fields are just const currently? (In nightly)

**Comment by jhpratt** [comment]
@Phosphorus-M Correct: https://github.com/rust-lang/rfcs/blob/master/text/3681-default-field-values.md#default-fields-values-are-const-contexts

**Comment by pk5ls20** [comment]
While using this feature, I encountered an internal rustc error: https://github.com/rust-lang/rust/issues/137136, and I checked the feature tracker but didn't find any similar issues.

**Comment by bstrie** [comment]
It appears as though the list of items for implementation is complete, and I see no blocking unresolved questions. If I submitted a PR to resolve the remaining documentation requirements, would we be ready for a stabilization PR?

**Comment by crumblingstatue** [comment]
> Expand support to non-const values? (Personal position is that we shouldn't do that, particularly seeing how powerful const eval is becoming.)

I personally have a couple of places where I have non-const values for defaults.
Example:
```Rust
#[derive(SmartDefault)]
pub struct Input {
    down: FnvHashSet<Key>,
    pressed: FnvHashSet<Key>,
    #[default(default_key_bindings())]
    pub key_bindings: FnvHashMap<InputAction, Key>,
    ...
}

fn default_key_bindings() -> FnvHashMap<InputAction, Key> {
    FnvHashMap::from_iter([
        (InputAction::Down, Key::S),
        (InputAction::Left, Key::A),
        (InputAction::Up, Key::W),
        (InputAction::Right, Key::D),
        (InputAction::Jump, Key::Space),
        ...
    ])
}
```

If default field values won't support non-const values, that means there will still be a place for crates like [`smart_default`](<https://docs.rs/smart-default/latest/smart_default/derive.SmartDefault.html>) in the ecosystem, which is fine. Although in that case I might as well use `smart_default` for everything rather than mix and match the two in a codebase.

If const traits (`from_iter`) and const alloc (`HashMap` needs to allocate) are expected to become a thing in the not too distant future, then it's not a big point of contention.

**Comment by estebank** [comment]
> It appears as though the list of items for implementation is complete, and I see no blocking unresolved questions.

@bstrie when adopting the feature in the rustc codebase, I encountered one hold-up: syn understandably doesn't understand this new feature, and we [can't add the trivial support](https://github.com/dtolnay/syn/pull/1851) because it'd be a breaking change. There's a way to add support for this on the stable syn version, but I haven't had the time to learn how to do that. After doing that, I'd have to go over every `derive` of types that I'd want to change and change their implementation if needed (most wouldn't, some might). If this is a concern for rustc, it'll be a concern for our users as well, so ideally we'd go around adding support to as much of the ecosystem as possible on the run-up to stabilization. This is not *critical*, but it'd be ideal to preclude people getting a sour taste from a feature they want to use but have to wait for for reasons outside their control. It'd feel like bait and switch. If you want to help, filling a syn PR with the approach dtolnay requested and going over your own codebases to see if there are third-party `derive`s that would need to be changed would be incredibly helpful. Also, getting experience reports from nightly crates, as well as any bug reports that come from that is also useful.

> If default field values won't support non-const values, that means there will still be a place for crates like [smart_default](https://docs.rs/smart-default/latest/smart_default/derive.SmartDefault.html) in the ecosystem, which is fine.

+1. I think that's fine.

> Although in that case I might as well use smart_default for everything rather than mix and match the two in a codebase.

That's fair.

> If const trait impls (`from_iter`) and const alloc (`HashMap` needs to allocate) are expected to become a thing in the not too distant future, then it's not a big point of contention.

I'm pretty sure these are coming, which changed my opinion on whether we'd potentially open this to non-const. Also remember that restricting to `const`s is not a one way door, we could expand support in the future if the ecosystem does find this too restrictive in practice.

**Comment by dhardy** [comment]
FYI `impl-tools` has supported [default field initialisers](https://crates.io/crates/impl-tools#impl-default) for some time, but requires quite a bit of macro code (including variants of several `syn` types) to do so. I look forward to this replacement (except that `derive` still doesn't get generic bounds right, but that's OT).

**Comment by zopsicle** [comment]
Should this feature be usable with unions? Currently union fields can be defined with default values but there appears to be no way to access them. I created a bug report for this: https://github.com/rust-lang/rust/issues/142555

**Comment by jhpratt** [comment]
Definitely not. For macros I think it would make sense to syntactically accept it, later rejecting it semantically.

**Comment by camsteffen** [comment]
Usage report.

I just tried out this feature in my project and it is awesome! I was able to tear out a custom builder pattern and replace it with struct literals which is so much cleaner.

Not having const Default feels like a pretty notable gap. If I have an enum with a `#[default]` variant, I am forced to repeat that default explicitly for field defaults instead of just writing `= default()`. Sometimes explicit is good, but oftentimes I want to encapsulate the default. This complicates the decision of switching to the new pattern.

**Comment by nayeemrmn** [comment]
Will there potentially be some way of defining implied field defaults for types? E.g.

```rust
// Built-in
trait ConstDefault {
  const CONST_DEFAULT: Self;
}

// Built-in
impl<T> ConstDefault for Option<T> {
  const CONST_DEFAULT: Self = None;
}

struct Flags {
  foo: Option<Foo>, // implicit `= Option::<Foo>::CONST_DEFAULT`
  bar: Option<Bar>, // implicit `= Option::<Bar>::CONST_DEFAULT`
}

let flags = Flags { .. }
```

**Comment by estebank** [comment]
> I just tried out this feature in my project and it is awesome! I was able to tear out a custom builder pattern and replace it with struct literals which is so much cleaner.

<3

> Not having const `Default` feels like a pretty notable gap. If I have an enum with a `#[default]` variant, I am forced to repeat that default explicitly for field defaults instead of just writing `= default()`. Sometimes explicit is good, but oftentimes I want to encapsulate the default. This complicates the decision of switching to the new pattern.

You'll be happy to see that it's going to be relatively straightforward to have exactly that, at least for the stdlib: https://github.com/rust-lang/rust/pull/134628

> Will there potentially be some way of defining implied field defaults for types?

No. That would have to be written as follows:

```rust
struct Flags {
  foo: Option<Foo> = Option::<Foo>::CONST_DEFAULT,
  bar: Option<Bar> = Option::<Bar>::CONST_DEFAULT,
}
```

**Comment by scottmcm** [comment]
...and, with other things in the works, just
```rust
use Default::default;
struct Flags {
  foo: Option<Foo> = default(),
  bar: Option<Bar> = default(),
}
```

(Completely implicit will not happen, because it's critical that type authors can *not* have a default, especially if that `Default` impl didn't exist for the field type when the outer type was written, but got added later.)

**Comment by recatek** [comment]
It would be nice to have some sort of opt-in implicit behavior for `= Default::default()`. If you're generating, say, a builder struct via macro, the macro can't know whether or not each field's type implements `Default`. As such, even with this feature implemented, there's still no way to do the following:

```rust
pub struct NoDefault;

pub struct Mixed {
    a: u32,
    b: NoDefault,
    c: Option<NoDefault>,
}

fn foo() -> Mixed {
    Mixed {
        b: NoDefault,
        ..Default::default()
    }
}
```

That is, allow me to do what I actually want to do with FRU here which is to explicitly specify only the fields that can't be defaulted, and fill in the rest implicitly.

**Comment by estebank** [comment]
@recatek As of tomorrow's nightly, [you will be able to describe that](https://github.com/rust-lang/rust/pull/134628) as

```rust
pub struct NoDefault;

pub struct Mixed {
    a: u32 = 0,
    b: NoDefault,
    c: Option<NoDefault> = Default::deafult(),
}

fn foo() -> Mixed {
    Mixed {
        b: NoDefault,
        ..
    }
}
```

A potential future extension of the feature is to make `_` mean `Default::default()` so that you can write the above as

```rust
pub struct Mixed {
    a: u32 = _,
    b: NoDefault,
    c: Option<NoDefault> = _,
}
```

**Comment by recatek** [comment]
Don't want to derail too much here if there's another discussion about using placeholders, but would `_` work for types that _don't_ implement `Default` in this case? Thinking again about macros that generate structs for the builder pattern but can't tell whether or not a given type is `Default`. In effect I'd like a way to say "as a macro, I don't know if this field is `Default`. If it is, use `Default::default()`, otherwise, require it to be specified during construction instead of eliding it via FRU".

**Comment by recatek** [comment]
Follow up on my earlier comment, and again I hope I'm not derailing, but I tried using "poor man's specialization" to work around my issue above (where the builder-generating macro doesn't know whether or not a type impls `Default`): https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=be7ee0b4f385e0c3c717922102cddba3

It looks like the default is still eagerly evaluated even if it's replaced in the FRU on initialization, or at least that's my guess here. Is that the ultimate intended behavior?

**Comment by kouhe3** [comment]
rust-analyzer 1.91.0-nightly (07d246fc 2025-08-31) incorrectly reports E0063 for the following code:

```rust
#![feature(default_field_values)]
#[derive(Debug, Default)]
struct Person<'a> {
    name: &'a str = "Default",
    age: u8,
}
fn main() {
    println!("{:?}", Person { age: 20, ..  });
}

```
rust-analyzer :
missing structure fields: - name

but code compiles successfully with the latest nightly compiler.

**Comment by jhpratt** [comment]
@kouhe3 That should presumably go in the r-a repo.

**Comment by agraven** [comment]
I also did a PR for adding a mention in The Book https://github.com/rust-lang/book/pull/4589

**Comment by DJMcNab** [comment]
dtolnay/syn#1870 is extremely painful to use, unfortunately.
Because it basically requires that anywhere you use a `Type`, you have to handle the "verbatim" case. That's not so bad if you're pulling it out into your own structure, but the problem arises when you're also passing around e.g.  a `Struct`, which is common (at least, it was the issue in clap).

I can see two ways forward here:

- David releases `syn` v3.0 with https://github.com/dtolnay/syn/pull/1851. This would be my preferred way forward
- David adds a `FieldWithDefault`, `StructWithDefault`, etc. to `syn`. This is obviously ugly, but at least it is suitably ergonomic enough for macros to actually want to port to it, and would have a smooth off-ramp once syn v0.3 releases.

However, both of these basically require David to decide that this feature is now "closer to being stable" enough to be supported in `syn`. The strongest signal in that regard is his lack of engagement with dtolnay/syn#1870, which implements a solution based on the principles he proposed in dtolnay/syn#1851 (but admittedly with changes for practicality of implementation). Based on this, I don't think that blocking on David is a viable path forward to get this stabilised.

My proposed path forward is for *someone* to open a pr against syn which implements solution 2, and then start the stabilisation process here (once the other blockers are resolved). I presume that we could schedule a future stabilisation, i.e. if we do a successful FCP, we can say "we'll stabilise this on nightly in 12 weeks/for the beta in 13-18 weeks". That would give David at least 19 weeks to decide between those paths forward, and the rest of that time for macros to port to the new versions.

**Comment by dhardy** [comment]
@marziply this already is the behaviour of `#[derive(Default)]`; (try using the existing `..Default::default()` syntax).

My concern is that the `..` syntax may cause confusing messages:
```
error: missing field `b` in initializer
  --> src/main.rs:11:39
   |
11 |     let s = S { c: "abc".to_string(), .. };
   |                                       ^ fields that do not have a defaulted value must be provided explicitly
```
It would be clearer if this suggested use of `..Default::default()` as an alternative.

Or maybe `..Default::default()` should be deprecated? In any case, I would suggest stabilising support for default field values without the new `..` syntax as a first step.

**Comment by theemathas** [comment]
Default field values use the same `Thing { field: value, .. }` syntax as rest-patterns in assignee expressions. This potentially complicates specifying the syntax in the reference, since it would need to specify a disambiguation rule for when an expression is using default field values, and when it's using a rest-pattern. See https://github.com/rust-lang/reference/issues/2208

**Comment by estebank** [comment]
@ronnodas [This lint already exists: `default_overrides_default_fields`](https://github.com/rust-lang/rust/pull/134737).

@FeldrinH note this was included in the RFC in the [drawbacks](https://github.com/rust-lang/rfcs/blob/master/text/3681-default-field-values.md#drawbacks) and Rationale ([on `const` contexts](https://github.com/rust-lang/rfcs/blob/master/text/3681-default-field-values.md#on-const-contexts) and [Instead of `Foo { ..Default::default() }`](https://github.com/rust-lang/rfcs/blob/master/text/3681-default-field-values.md#instead-of-foo--defaultdefault-)) sections. `Foo { .. }` can do more than `Foo { ..Default::default() }` can (*even* if `~const Default::default()` was stable). There's a need for its *functionality*.

The comment you mention seems to have been more concerned with the verbosity of adopting `..` to avoid `..Default::default()`. The RFC mentions the possibility of using `_` or another marker for "delegate to the `~const Default` for this field's type", making it easier to deal with. For reference, in `rustc` [we use default field values already](https://github.com/rust-lang/rust/blob/main/compiler/rustc_resolve/src/lib.rs#L1163-L1382), and I haven't heard complaints from other compiler contributors.

**Comment by abellgithub** [comment]
When does `Foo { .. }` work?

The following doesn't compile:

```
#[derive(Default)]
struct Foo
{
    i1: i32 = 0,
    i2: i64 = 1,
}

fn main()
{
  let foo = Foo { .. };
}
```

**Comment by rodrimati1992** [comment]
> When does `Foo { .. }` work?
> 
> The following doesn't compile:
> 
> ```
> #[derive(Default)]
> struct Foo
> {
>     i1: i32 = 0,
>     i2: i64 = 1,
> }
> 
> fn main()
> {
>   let foo = Foo { .. };
> }
> ```

The code doesn't work because the unstable feature for default field values isn't enabled, it needs this to compile:
```diff
+ #![feature(default_field_values)]
```
working version: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=352d493714934883ba0c7ba6ab14a90e

**Comment by maia-s** [comment]
If you try to use `..` on a struct with no fields, you get an error message that the `..` syntax requires a struct with at least one field. I think it'd be nice if it worked on structs with no fields too, so that they can be expanded later without breaking API if `..` is used.

I also don't understand why `#[non_exhaustive]` isn't allowed on these structs, for the same reason.

I didn't see any explanation above for why these are disallowed, only statements that they are

**Comment by jhpratt** [comment]
@maia-s The RFC has [a section explaining why](https://rust-lang.github.io/rfcs/3681-default-field-values.html#non_exhaustive-interactions). Essentially, a field can be added in the future that is not defaulted, and this must remain a non-breaking change. Alternatives are listed.

**Comment by ennis** [comment]
> [@maia-s](https://github.com/maia-s) The RFC has [a section explaining why](https://rust-lang.github.io/rfcs/3681-default-field-values.html#non_exhaustive-interactions). Essentially, a field can be added in the future that is not defaulted, and this must remain a non-breaking change. Alternatives are listed.

I encountered the same error `.. needs at least one default field in the struct definition` while trying to default-initialize an empty struct. The linked section of the RFC deals with `#[non_exhaustive]`, and I think I understand the reason for that. However, I don't understand why initializing empty structs _without_ `#[non_exhaustive]` are disallowed (e.g.: `S { .. }` for `struct S {}`). The section about [“Empty” types and types without default field values](https://rust-lang.github.io/rfcs/3681-default-field-values.html#empty-types-and-types-without-default-field-values) mentions:

> I propose we disallow this at least initially. .. can then only be used if there is at least one default field. We might want to change this rule in the future, but careful with how it would interact with #[non_exhaustive], as it could accidentally allow for types that are not meant to be constructed outside of a given crate to all of a sudden be constructable.

Which mentions `#[non_exhaustive]` again. What problems would allowing `S { .. }` for empty structs (without `#[non_exhaustive]`) cause?

**Comment by jhpratt** [comment]
@ennis An empty struct as in `struct Foo {}` or `struct Foo;`? For the former, I agree: there is no technical reason to prohibit it. For `struct Foo;`, I _thought_ that you couldn't match on that with `{ .. }`, but a quick check shows you can. So I guess there's no reason to prohibit it either way.

@estebank thoughts? This seems like a straightforward decision imo

**Comment by ennis** [comment]
> [@ennis](https://github.com/ennis) An empty struct as in `struct Foo {}` or `struct Foo;`? For the former, I agree: there is no technical reason to prohibit it. For `struct Foo;`, I _thought_ that you couldn't match on that with `{ .. }`, but a quick check shows you can. So I guess there's no reason to prohibit it either way.

In my case, it was an empty struct as in `struct Foo {}` (a non-unit struct without fields)

**Comment by agraven** [comment]
I would suggest adding https://github.com/dtolnay/syn/pull/1955 as a sub-bullet to the `Support parsing in syn` bullet point

**Comment by eirnym** [comment]
I love the overall idea and want to clear few questions on which RFC don't answer. 

---

RFC doesn't clearly say what default value could be: literal or any const expression. 

E.g. would be I allowed to write snippet below

```rust
struct Foo {
   pub field: &'static str = bar();
}

const fn bar() -> &'static str {}

fn creator() {
    let foo = Foo { .. };
}
```

---

What is visibility context for this: module of definition or call site. 

Could you please also explain if feature would be implemented now with caller context, how easy (or breaking) it would be for existing code to change context to definition site?

**Comment by mumbleskates** [comment]
i'm sure it would be extremely confusing and dangerous for the definition to apply at the callsite. i would certainly expect that this was always planned and implemented to resolve the default expression at the definition.

**Comment by ds84182** [comment]
Is the initialization of defaults guaranteed to occur after all other field initializers? This currently compiles (as expected), but could stop compiling if x or i2 are initialized before the break within i1 since String cannot be dropped within const.

```rust
#![feature(default_field_values)]

#[derive(Default)]
struct Foo {
    x: String = String::new(),
    i1: i32 = 0,
    i2: String = String::new(),
}

fn main() {
    let foo = const { 'out: { Foo { i1: break 'out Foo { .. }, .. } } };
    println!("{} {}", foo.i1, foo.i2);
}
```

**Comment by cart** [comment]
Default field values are _exactly_ what we need over in Bevy land! (here is some [Discord context](https://discord.com/channels/691052431525675048/749332104487108618/1524159485952921790) and [an intended usage example that compares against the alternative](https://discord.com/channels/691052431525675048/1476701089746255912/1524563690014248991) for those interested in the details).

In addition to the obvious and very desirable ergonomic benefits, "default field values" are perfect for expressing "partial defaults" (for types where some fields can have defaults and others cant), and they would allow us to share default values between a `Component` and its derived / generated  `Template` type (as the Template derive would have access to the default field values).

This feature is at the top of my personal wishlist for Bevy, as it would allow us to embrace field-driven APIs ergonomically, while introducing some types of fields that cannot be default (notably, always-strong `Handle`s in our asset system). This is an incredibly important part of the puzzle for us (and our users).  We likely can't (or at least won't) move in the direction we want to until this lands, as "partial defaults" are otherwise too much of a UX issue without them.

We are heavily invested in this landing. The sooner it lands the better, and I'm happy to spend effort to unblock any areas that need help. I think the current implementation is in a solid spot. I wouldn't personally block this on `syn` support. Better to make the progress we can, when we can. `syn` support will naturally (and likely quickly) spawn from the feature landing.

That being said there is one issue that I think needs serious consideration (ideally before this lands, but it seems like it could be added later):

> Expand support to non-const values? (Personal position is that we shouldn't do that, particularly seeing how powerful const eval is becoming.)

I emphatically think non-const values should be supported.

By only supporting const values, this creates _significant_ seemingly arbitrary restrictions on who and what can use this feature. The reality of the situation is that the Rust ecosystem has a large (majority?) body of  non-const code, both for historical limitations (const hasn't always been as capable as it is today), social reasons (functions require explicit const labels), current restrictions (const still isn't capable enough to express many things), and technical reasons (some things literally cant be const).

The biggest and most painful example of this is obviously `Default`. I immediately hit this restriction when trying to port Bevy code (ex:  `struct Foo { bar: Bar = Default::default() }`). This is THE natural thing to want. We're defining "_default_ field values"!  _Of course_ the natural default for `Bar` is going to be `Bar::default()`. Yes, const Default is planned, but that may or may not land before this feature (and it might land many years later for all we know). And it will likely require explicit opt-in (meaning the entire current body of `Default` in the ecosystem will be unusable with field value defaults). And of course, there are many examples of Default that _cant_ be `const`, even if that feature lands. 

There are other obvious examples like `vec![ A, B ]` and `String::new("hello")`.  Imagine all of the UI use cases that are being abandoned because you can't do things like this:

```rust
struct Button {
  text: String = String::new("Button"),
}
```

Imagine the frustration and toil that will occur when someone builds a "field defaults" API (that _happens_ to be const). Your users adopt that API en-mass. Later, you learn that you need to add a field whose default cannot be const, and suddenly you (and all of yours users) need to break and adopt a different (likely much worse) coding pattern because `Foo { .. }` can no longer accommodate your needs.

The rationale for the const requirement in the RFC seems to boil down to:
- **Determinism**: allowing non-const means that this could happen `Foo {..} != Foo {..}`. I would argue that this is fine, and it is the API author's right to build something like that. We don't balk at `Foo::new() != Foo::new()` or `Foo::default() != Foo::default()`. There are plenty of perfectly valid reasons to want to build something like this:
  ```rust
  struct WorldGenerator { seed: usize = rand() }
  ```
- **Cheapness**: Yes const is more likely to be "cheaper". But I would argue that it is the type author's right to decide how "expensive" their types should be to construct. Different contexts demand different initialization "costs". And notably, many, many contexts _cant_ be initialized via `const`. "Initialization ergonomics" and "partial defaults" should not be held hostage because a type might be too expensive to initialize.

There is also notably _no_ pre-existing "cheapness" or determinism expectation here for Rust, given that `Foo { ..Default::default() }` is already "the" way to initialize Foo with default values. 

We'd love to use `Foo { .. }` pretty much everywhere in Bevy. I think the Rust ecosystem _in general_ would love to use it everywhere. That can't happen if `const` is required. People will hit this restriction and (like so many times before) be forced to make their code worse for seemingly arbitrary reasons.

**Comment by scottmcm** [comment]
On `const`-ness: I think a core question is how prevalent the cases are that *can't* be `const` vs the cases that *currently* aren't `const` but where that's already an obvious wish for other cases too.

Notably there's a project goal for const traits already (https://rust-lang.github.io/rust-project-goals/2026/const-traits.html) 

> The compiler already has a mature implementation of const traits. We will draft the RFC to address any outstanding concerns from existing feedback. Afterwards we will call for testing and push for stabilization.

so I'd expect it's not "many years later for all we know".

---

I'd like to hear more about specifically the cases where you think it *can't* be `const`.  How should I weigh "well I want a `String`" vs "it'd be const if you used `Cow<'static, str>`"?  It makes me think of `static`s, where we *could* have said "it's the author's right to do weird things before main", but intentionally don't.  How does `seed: usize = rand()` compare to trivially-`const` things like `seed: Option<usize> = None` or `seed: Lazy<usize> = Lazy::new(rand)`?

---

Also, what's your perspective on breakage from adding fields?  Are you fine with that because bevy revs the major version frequently, or?  I would have expected that to be a bigger stumbling block than `const`-ness, TBH.

**Comment by cart** [comment]
> so I'd expect it's not "many years later for all we know".

This is good to hear!

> I think a core question is how prevalent the cases are that can't be const vs the cases that currently aren't const but where that's already an obvious wish for other cases too

I think thats a relevant question, but I wouldn't consider it to be a deciding one. None of the planned const efforts solve the problem of "many things cant be const" (allocations, global context, etc) and "every single trait / impl currently in the Rust ecosystem is not const, as are many (most?) of the functions someone might want to use to initialize a type". Const traits, which are opt-in, don't solve the latter problem.

>  It makes me think of statics, where we could have said "it's the author's right to do weird things before main", but intentionally don't.

I think this is meaningfully different, provided `Foo { .. }` is thought of more like `..default()` instead of a "a piece of memory prepared by the compiler ahead of time and copied from when needed". This isn't meaningfully "life before main" levels of weirdness. Just a normal initialization logic with nicer syntax than we've had before.

I'd like to understand the pressing need to _force_ `const` (which I will define as BENEFITS_OF_FORCED_CONST so I can reference them later). Why can't it just be const (and get all of the benefits of that) if all of the field initializers are also const? What does the forced constraint win the average user that offsets the _massive_ usability cost (and ensuing developer frustration)? 

> How should I weigh "well I want a String" vs "it'd be const if you used Cow<'static, str>"?

I think you should weigh "you can't use a significant portion of Rust types, both in the standard library and in the ecosystem" quite highly personally. An important subset of that category is "things that can expand over time" (which is a _massive_ category including things like lists/vec, graphs, strings, etc). It doesn't seem reasonable to restrict something as central as "the preferred / most ergonomic initialization syntax" to a space that doesn't include "dynamically changing things".

The Cow vs String case has a number of (minor but notable) tradeoffs, that each individual type author should be free to make.
- The obvious: `String` is easier to type / nicer to look at / the first string type developers often reach for (especially the new ones)
- `Cow<'static, str>` is functionally different than `String`. `String` has methods like `value.push('w')`.  This is asking the developer to fundamentally change their API's interface. They may have chosen `String` because that is the string type that serves their purposes the best. Are we going to force them to use `Cow<'static, str>` instead in service of BENEFITS_OF_FORCED_CONST?  
- Most functions that return a generated string return a `String` (including things like `format!()`). It means that cases that would otherwise be `my_type.foo = format!("x {y}");` now need to be `my_type.foo = format!("x {y}").into()`. This degrades ergonomics.
- Currently, Rust has a perception that "there are too many string types" (I don't personally share this sentiment, but I do understand it). Do we really want to add to the list of "string type frustrations" that String can't be used in default field initializers? (the thing that I expect most people to be defining their defaults in going forward)

I consider all of the issues above to be "important" from a language design perspective. And notably, none of them are specific to Cow vs String. These are global issues that extend into everything (for some reasonable definition of "everything" 😄 ) 

> How does seed: usize = rand() compare to trivially-const things like seed: Option<usize> = None or seed: Lazy<usize> = Lazy::new(rand)?

This is essentially asking "are you willing to make compromises on the following" to get BENEFITS_OF_FORCED_CONST:
- **Less ergonomic initialization**: `Foo { seed: Some(10) }` vs `Foo { seed: 10 }`
- **Less performant runtime api**: Unwrapping at runtime is necessary (`foo.seed.unwrap()` vs `foo.seed`)
- **Less ergonomic consumption api**: the developer is now forced to contend with getting the type out of the `Option`. Same with `Lazy` (ex: `forced_mut()` instead of just modifying the type directly).
- **Higher type complexity**: `Option<usize>` / `Lazy<usize>` / etc are all just one more "thing" that someone reading / using the code needs to think about 

Given that the BENEFITS_OF_FORCED_CONST are extremely unclear to me (especially when compared to const-just-when-all-field-initializers are const), my answer is going to default to "I would absolutely not make that tradeoff" personally, and I expect most users to feel the same way.

> Also, what's your perspective on breakage from adding fields? 

I'm ok with this. Functionally, adding a field without a default field value is "like" adding another arguments to a constructor (ex: `Foo::new(x)` becomes `Foo::new(x, y)`). That would be a breaking change.

Notably, isn't adding a non-default field being a breaking change already the case?

**Comment by RoloEdits** [comment]
To add some small dialog on this, using syntax from a proposed RFC https://github.com/rust-lang/rfcs/pull/3444, where the ergonomic changes could really help drive towards other construction patterns besides just a builder.

Using the bevy example above:

```rust
struct WorldGenerator { name: String,  seed: usize = rand() }
let name: String = name_from_user_input();
let world = World::new(.{ name , .. });
```

So rather than having to have a `WorldBuilder` struct, which would need a `builder.name(name)`, we get configuration struct pattern, where the fields act like named arguments, but this really only works best of you can construct the values from any source, not just known:

```rust
let connection = Connection::open(.Https {
    host: "localhost",
    port: 5672,
    username: "user",
    password: "bitnami",
    ..
})
.await;
```

**Comment by vultix** [comment]
> How does seed: usize = rand() compare to trivially-const things like seed: Option = None or seed: Lazy = Lazy::new(rand)?

One other important thing to call out here. `seed: Option<usize> = None` has a semantically different meaning than `seed: usize = rand()`. The former implies that the seed is optional, when in fact the seed _is_ required. Initializing the seed is the optional part - the type definition should reflect this invariant. 

You could add a comment documenting "None means random," but that papers over a type that lies about the invariant.

**Comment by agraven** [comment]
Syn support for default values has been added in 3.0.0, specifically by this PR https://github.com/dtolnay/syn/pull/2054
